diff --git a/src/java/org/apache/nutch/crawl/DbUpdateReducer.java b/src/java/org/apache/nutch/crawl/DbUpdateReducer.java
index 16ba2db..683d480 100644
--- a/src/java/org/apache/nutch/crawl/DbUpdateReducer.java
+++ b/src/java/org/apache/nutch/crawl/DbUpdateReducer.java
@@ -183,9 +183,10 @@ extends GoraReducer<UrlWithScore, NutchWritable, String, WebPage> {
     }
     Mark.GENERATE_MARK.removeMarkIfExist(page);
     Mark.FETCH_MARK.removeMarkIfExist(page);
-    Utf8 mark = Mark.PARSE_MARK.removeMarkIfExist(page);
-    if (mark != null) {
-      Mark.UPDATEDB_MARK.putMark(page, mark);
+    Utf8 parse_mark = Mark.PARSE_MARK.checkMark(page);
+    if (parse_mark != null) {
+      Mark.UPDATEDB_MARK.putMark(page, parse_mark);
+      Mark.PARSE_MARK.removeMark(page);
     }
 
     context.write(keyUrl, page);
